### PR TITLE
fix: hide priority field for profiles without priority change rights

### PR DESCRIPTION
**Pull request by ccailly**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #13603 

The priority field in the ticket form was visible to all users. It was disabled, for those who did not have the right to modify the priority. Except that the priority was "reset" when changing category, for example. For more information, please refer to issue #13603. The changes made in this RP ensure that the priority field is hidden for these users, but remains visible to users with the appropriate rights.

### CODE REVIEW

**Comment by Rom1-B** [comment]
I don't know if this is the right approach? Because without the right to modify the priority, we can't even see the read-only value?

@cedric-anne @orthagh

**Comment by cedric-anne** [comment]
> I don't know if this is the right approach? Because without the right to modify the priority, we can't even see the read-only value?
> 
> @cedric-anne @orthagh

I guess we could keep this information displayed, and rely on the `computePriority` method. The following code is computing the default priority https://github.com/glpi-project/glpi/blob/329b50bd6a9f1b5c96c035bd294d23484f7a4cc2/src/CommonITILObject.php#L803-L815 but is only working on first display. I think it could be fixed to handle the reload case (not sure, because this method is really hard to understant, due to a mix of $options, $this->fields and $predefined_fields handling at the same time).

File: src/Ticket.php [hunk 1]
```diff
@@ -4409,6 +4409,14 @@ public function showForm($ID, array $options = [])
...
         }
 
+        // compute priority if is new ticket and user have no right to change priority
+        if ($this->isNewItem() && !$canpriority) {
+            $this->fields['priority'] = $this->computePriority(
+                $this->fields['urgency'] ?? 3,
+                $this->fields['impact'] ?? 3
+            );
+        }
+
```
Comment (Rom1-B) [inline src/Ticket.php:4419]: Maybe something like this?

```suggestion
        // compute priority if is new ticket and user have no right to change priority
        if ($this->isNewItem() && !$canpriority) {
            $this->fields['priority'] = $this->computePriority(
                $this->fields['urgency'] ?: $predefined_fields['urgency'] ?? 3 /* Medium */,
                $this->fields['impact'] ?: $predefined_fields['impact']  ?? 3 /* Medium */
            );
        }

```

File: src/Ticket.php [hunk 2]
```diff
@@ -4409,6 +4409,14 @@ public function showForm($ID, array $options = [])
             $options['_noupdate'] = true;
         }
 
+        // compute priority if is new ticket and user have no right to change priority
+        if ($this->isNewItem() && !$canpriority) {
+            $this->fields['priority'] = $this->computePriority(
+                $this->fields['urgency'] ?: $predefined_fields['urgency'] ?? 3 /* Medium */,
+                $this->fields['impact'] ?: $predefined_fields['impact']  ?? 3 /* Medium */
+            );
+        }
```
Comment (cedric-anne) [inline src/Ticket.php:4418]: This seems a bit redundant with the code present in `setPredefinedFields()`: https://github.com/glpi-project/glpi/blob/329b50bd6a9f1b5c96c035bd294d23484f7a4cc2/src/CommonITILObject.php#L803-L815

The same kind of code will be executed twice when rendering the ticket form. I guess there is something that could be simplified here (but, again, this code is a mess, so I am not sure of anything).

**Comment by cedric-anne** [comment]
I close this PR for now. Fixing the issue seems more complex than expected and we already spend too much time for a bug that is, IMHO, not so important.

We may try to solve this issue later, but for now, we have to focus on more important topics.

**Pull request closed without merging**